storage/shard.go

```diff
@@ -166,6 +166,10 @@ func (s *dbShard) QueryRaw(
 	if err != nil {
 		return nil, err
 	}
+	if res == nil {
```

Comment:
Thanks fo the catch. I feel fhis should be done inside the `QueryRaw` method of the mutable segment, i.e., maintaining the invariant that if there is no error, the result is always non nil.